src/database.ts

```diff
@@ -0,0 +1,100 @@
+import { type Result } from "./registry/index.js";
+import {
+  type Config,
+  type AutoWaitConfig,
+  extractBaseUrl,
+  type ChainName,
+  getBaseUrl,
+  type Signal,
+  type Signer,
+  normalize,
+} from "./helpers/index.js";
+import { Statement } from "./statement.js";
+
+export class Database<D = unknown> {
+  readonly config: Config & Partial<AutoWaitConfig>;
+
+  constructor(config: Config & Partial<AutoWaitConfig> = {}) {
+    this.config = config;
+  }
+
+  static async readOnly(chainNameOrId: ChainName | number): Promise<Database> {
+    const baseUrl = getBaseUrl(chainNameOrId);
+    return new Database({ baseUrl });
+  }
+
+  static async forSigner(signer: Signer): Promise<Database> {
```

Comment:
This will determine which baseUrl to use given the chain that this signer is connected to. These two static methods are probably the best ways for devs to construct a new Database instance.